src/stream/src/executor/sync_kv_log_store.rs

```diff
@@ -535,38 +535,40 @@ impl<S: StateStore> SyncedKvLogStoreExecutor<S> {
         truncation_offset: Option<ReaderTruncationOffsetType>,
         buffer: &mut SyncedLogStoreBuffer,
     ) -> StreamExecutorResult<LogStorePostSealCurrentEpoch<'a, S::Local>> {
-        let epoch = barrier.epoch.prev;
-        let mut writer = write_state.start_writer(false);
-        writer.write_barrier(epoch, barrier.is_checkpoint())?;
+        // TODO(kwannoel): As an optimization we can also change flushed chunks to be flushed items
+        // to reduce memory consumption of logstore.
 
-        // FIXME(kwannoel): Flush all unflushed chunks
-        // As an optimization we can also change it into flushed items instead.
-        // This will reduce memory consumption of logstore.
+        let epoch = barrier.epoch.prev;
+        if barrier.is_checkpoint() {
+            let mut writer = write_state.start_writer(false);
+            writer.write_barrier(epoch, barrier.is_checkpoint())?;
 
-        // TODO: may stop the for loop when seeing any of flushed item to avoid always iterating the whole buffer
-        for (epoch, item) in &mut buffer.buffer {
-            match item {
-                LogStoreBufferItem::StreamChunk {
-                    chunk,
-                    start_seq_id,
-                    end_seq_id,
-                    flushed,
-                    ..
-                } => {
-                    if !*flushed {
-                        writer.write_chunk(chunk, *epoch, *start_seq_id, *end_seq_id)?;
-                        *flushed = true;
+            for (epoch, item) in &mut buffer.buffer {
+                match item {
+                    LogStoreBufferItem::StreamChunk {
+                        chunk,
+                        start_seq_id,
+                        end_seq_id,
+                        flushed,
+                        ..
+                    } => {
+                        if !*flushed {
+                            writer.write_chunk(chunk, *epoch, *start_seq_id, *end_seq_id)?;
+                            *flushed = true;
+                        } else {
+                            break;
+                        }
                     }
+                    LogStoreBufferItem::Flushed { .. } => {
```

Comment:
Not sure if we can break when seeing `Flushed` item. For example, let's say the buffer size is 1 chunk, and then we have the following actions

start from empty buffer: []
add chunk 1: [chunk1] 
add flushed chunk2: [chunk1, flushed2] 
consume chunk1: [flushed2]
add chunk2: [flushed2, chunk2]

On checkpoint barrier, when we should write all unflushed chunks, we will first see flushed2, and then break the for loop. But chunk2 is not flushed yet.

Reply:
Fixed